File: src/data/hullmods/aEP_MarkerDissipation.java
package data.hullmods;

import com.fs.starfarer.api.Global;
import com.fs.starfarer.api.combat.*;
import com.fs.starfarer.api.combat.ShipAPI.HullSize;
import com.fs.starfarer.api.combat.listeners.AdvanceableListener;
import com.fs.starfarer.api.combat.listeners.ApplyDamageResultAPI;
import com.fs.starfarer.api.combat.listeners.DamageListener;
import com.fs.starfarer.api.combat.listeners.DamageTakenModifier;
import com.fs.starfarer.api.ui.Alignment;
import com.fs.starfarer.api.ui.TooltipMakerAPI;
import com.fs.starfarer.api.util.Misc;
import combat.util.aEP_Tool;
import combat.util.aEP_DataTool;
import org.jetbrains.annotations.NotNull;
import org.lazywizard.lazylib.MathUtils;
import org.lwjgl.util.vector.Vector2f;

import java.awt.*;
import java.util.HashMap;
import java.util.Map;


import static data.hullmods.aEP_HotLoader.EXTRA_SPEED_ON_FIRE;
import static combat.util.aEP_DataTool.txt;

public class aEP_MarkerDissipation extends aEP_BaseHullMod
{


  public static final float BUFFER_AREA = 0.1f;//
  static final Map<HullSize, Float> WEAPON_RANGE_BONUS = new HashMap<>();
  static final float OVERLOAD_TIME_DECREASE = 0.25f;
  static final float DECREASE_SPEED = 0.5f;//
  static final float MAX_OVERLOAD_TIME = 7f;

  private static String id = "aEP_MarkerDissipation";


  boolean showNoModOnUp = false;


  public static float getBufferLevel(ShipAPI ship) {
    if (ship == null || !ship.isAlive() || !ship.getVariant().hasHullMod("aEP_MarkerDissipation")) return 0f;
    aEP_DataTool.floatDataRecorder fluxData = (aEP_DataTool.floatDataRecorder) Global.getCombatEngine().getCustomData().get(ship.getId()+"_"+id);
    if (fluxData == null) return 0f;
    float buffer = fluxData.getTotal();
    float maxFlux = ship.getFluxTracker().getMaxFlux();
    float bufferLevel = aEP_Tool.limitToTop(buffer / (maxFlux * BUFFER_AREA), 1f, 0f);
    //三次方会让曲线更快下降
    return MathUtils.clamp(bufferLevel * bufferLevel * bufferLevel * 2f, 0f, 1f);
  }

  @Override
  public void applyEffectsAfterShipCreationImpl(ShipAPI ship, String id) {
    this.id = id;
    //减少过载时间
    ship.getMutableStats().getOverloadTimeMod().modifyMult(id, 1f - OVERLOAD_TIME_DECREASE);
    //修改护盾贴图
    if (ship.getShield() != null) {
      //set shield inner, outer ring
      ship.getShield().setRadius(ship.getShield().getRadius(),
        Global.getSettings().getSpriteName("aEP_hullstyle", "aEP_shield_inner"),
        Global.getSettings().getSpriteName("aEP_hullstyle", "aEP_shield_outer"));
    }

    if (!ship.hasListenerOfClass(FluxRecorder.class)) {
      aEP_DataTool.floatDataRecorder fluxData = new aEP_DataTool.floatDataRecorder();
      ship.addListener(new FluxRecorder(ship, fluxData));
      Global.getCombatEngine().getCustomData().put(ship.getId()+""+id, fluxData);
    }
  }


  @Override
  public void advanceInCombat(@NotNull ShipAPI ship, float amount) {
    super.advanceInCombat(ship, amount);
    if(ship.getFluxTracker().isOverloaded() && ship.getFluxTracker().getOverloadTimeRemaining() > MAX_OVERLOAD_TIME){
      ship.getFluxTracker().setOverloadDuration(MAX_OVERLOAD_TIME);
    }
  }

  public String getDescriptionParam(int index, ShipAPI.HullSize hullSize) {
    return null;
  }

  public boolean shouldAddDescriptionToTooltip(HullSize hullSize, ShipAPI ship, boolean isForModSpec) {
    return true;
  }

  public void addPostDescriptionSection(TooltipMakerAPI tooltip, HullSize hullSize, ShipAPI ship, float width, boolean isForModSpec) {
    tooltip.addSectionHeading(txt("effect"), Alignment.MID, 5f);
    tooltip.addPara("- " + txt("overload_time_reduce") + "{%s}", 5f, Color.white, Color.green, (int) (OVERLOAD_TIME_DECREASE * 100) + "%");
    tooltip.addPara(aEP_DataTool.txt("MD_des04"), Color.gray, 5f);
  }


  void showMod03(TooltipMakerAPI tooltip, HullSize hullSize, ShipAPI ship, float width, boolean isForModSpec) {
    TooltipMakerAPI image = tooltip.beginImageWithText(Global.getSettings().getHullModSpec("aEP_HotLoader").getSpriteName(), 48f);
    image.addPara("- " + txt("ammo_regen_speed_up") + "{%s}", 5f, Color.white, Color.green, (int) (EXTRA_SPEED_ON_FIRE * 100) + "%");
    tooltip.addImageWithText(5f);
    showNoModOnUp = false;
  }



  class FluxRecorder implements DamageListener, DamageTakenModifier, AdvanceableListener
  {
    aEP_DataTool.floatDataRecorder fluxData;
    private ShipAPI ship;
    private Object param;
    private DamageAPI damage;

    public FluxRecorder(ShipAPI ship, aEP_DataTool.floatDataRecorder fluxData) {
      this.ship = ship;
      this.fluxData = fluxData;
    }

    @Override
    public void advance(float amount) {
      //get data
      float maxFlux = ship.getFluxTracker().getMaxFlux();
      float softFlux = ship.getFluxTracker().getCurrFlux() - ship.getFluxTracker().getHardFlux();
      float hardFlux = ship.getFluxTracker().getHardFlux();
      float totalDiss = ship.getMutableStats().getFluxDissipation().getModifiedValue();

      //renew after you already got the data from last frame
      fluxData.addRenewData(softFlux);

      //如果检测到本帧软幅能有所下降，开始消减buff区
      //limit soft flux data within buffer area and decrease buffer area
      fluxData.setTotal(aEP_Tool.limitToTop(fluxData.getTotal() - totalDiss * DECREASE_SPEED * amount, maxFlux * BUFFER_AREA, 0f));

      //control which string and boolean to show in stats
      float buffer = fluxData.getTotal();

      //venting check
      if (ship.getFluxTracker().isOverloadedOrVenting()) {
        fluxData.setTotal(0f);
      }
      float bufferLevel = aEP_Tool.limitToTop(buffer / (maxFlux * BUFFER_AREA), 1, 0);
      bufferLevel = MathUtils.clamp(bufferLevel * 2f, 0f, 1f);

      if (Global.getCombatEngine().getPlayerShip() == ship) {
        Global.getCombatEngine().maintainStatusForPlayerShip(this.getClass().getSimpleName(),//key
                "graphics/aEP_hullsys/marker_dissipation.png",//sprite name,full, must be registed in setting first
                Global.getSettings().getHullModSpec(id).getDisplayName(),//title
                txt("MD_des01") + (int) (bufferLevel * 100) + "%",//data
                false);//is debuff
      }
      Global.getCombatEngine().getCustomData().put(ship.getId()+"_"+id,fluxData);
    }

    //在 report之前被调用
    //用来记录 DamageAPI
    @Override
    public String modifyDamageTaken(Object param, CombatEntityAPI target, DamageAPI damage, Vector2f point, boolean shieldHit) {
      if(param == null) return null;
      this.param = param;
      this.damage = damage;
      return null;
    }

    @Override
    public void reportDamageApplied(Object source, CombatEntityAPI target, ApplyDamageResultAPI result) {
      //Global.getLogger(this.getClass()).info( "reportDamageApplied");
      //Global.getLogger(this.getClass()).info( result.getDamageToShields() + "");
      if(source == null || damage == null || result.getDamageToShields() <= 0) return;
      if ( (param instanceof BeamAPI && !damage.isForceHardFlux()) || damage.isSoftFlux()) {
        fluxData.setLastFrameData(fluxData.getLast() + result.getDamageToShields());
      }
      return;
    }
  }

}

File: src/data/hullmods/aEP_MarkerDissipation.kt
package data.hullmods

import com.fs.starfarer.api.Global
import com.fs.starfarer.api.combat.BeamAPI
import com.fs.starfarer.api.combat.CombatEntityAPI
import com.fs.starfarer.api.combat.DamageAPI
import com.fs.starfarer.api.combat.ShipAPI
import com.fs.starfarer.api.combat.ShipAPI.HullSize
import com.fs.starfarer.api.combat.listeners.AdvanceableListener
import com.fs.starfarer.api.combat.listeners.ApplyDamageResultAPI
import com.fs.starfarer.api.combat.listeners.DamageListener
import com.fs.starfarer.api.combat.listeners.DamageTakenModifier
import com.fs.starfarer.api.ui.Alignment
import com.fs.starfarer.api.ui.TooltipMakerAPI
import com.fs.starfarer.api.util.Misc
import combat.plugin.aEP_CombatEffectPlugin.Mod.addEffect
import combat.util.aEP_DataTool
import combat.util.aEP_DataTool.floatDataRecorder
import combat.util.aEP_DataTool.txt
import combat.util.aEP_ID
import combat.util.aEP_Tool
import data.scripts.weapons.aEP_TeslaBeam
import org.lazywizard.lazylib.MathUtils
import org.lwjgl.util.vector.Vector2f
import java.awt.Color

class aEP_MarkerDissipation : aEP_BaseHullMod() {

  companion object {
    const val BUFFER_AREA = 0.2f // 缓冲区总大小是幅能容量的百分之多少
    const val INCREASE_SPEED = 1f // 产生的幅能百分之多少能多少计入缓冲区
    const val DECREASE_SPEED = 0.5f // 缓冲区每秒下降幅能耗散的百分之多少
    const val DECREASE_PERCENT= 0.5f // 缓冲区每秒下降实际幅能降低的多少
    const val OVERLOAD_TIME_DECREASE = 0.25f
    const val MAX_OVERLOAD_TIME = 8f
    const val ID = "aEP_MarkerDissipation"
    @JvmStatic
    fun getBufferLevel(ship: ShipAPI?): Float {
      if (ship == null || !ship.isAlive || ship.isHulk || !ship.variant.hasHullMod(ID)) return 0f
      val fluxData = (ship.customData["$ID _ ${ship.id}"] ?: return 0f) as floatDataRecorder
      val buffer = fluxData.total
      val maxDissi = aEP_Tool.getRealDissipation(ship)
      val maxCap = ship.maxFlux
      val bufferLevel = buffer / (maxCap * BUFFER_AREA + 1f)
      //次方会让曲线更快下降
      return MathUtils.clamp(   bufferLevel * 2f, 0f, 1f)
    }
  }

  override fun applyEffectsAfterShipCreationImpl(ship: ShipAPI, id: String) {
    //减少过载时间
    ship.mutableStats.overloadTimeMod.modifyMult(ID, 1f - OVERLOAD_TIME_DECREASE)
    //修改护盾贴图
    if (ship.shield != null) {
      //set shield inner, outer ring
      ship.shield.setRadius(
        ship.shield.radius,
        Global.getSettings().getSpriteName("aEP_hullstyle", "aEP_shield_inner"),
        Global.getSettings().getSpriteName("aEP_hullstyle", "aEP_shield_outer")
      )
    }

    //当希望customData进行初始化时，条件一定要是customData.contains()，有的时候舰船生成到一半，custom此时还没有创建
    if (!ship.customData.containsKey("$ID _ ${ship.id}")) {
      val fluxData = floatDataRecorder()
      ship.setCustomData("$ID _ ${ship.id}",fluxData)
      ship.addListener(FluxRecorder(ship, fluxData))
    }

  }

  override fun advanceInCombat(ship: ShipAPI, amount: Float) {
    super.advanceInCombat(ship, amount)
    if (ship.fluxTracker.isOverloaded && ship.fluxTracker.overloadTimeRemaining > MAX_OVERLOAD_TIME) {
      ship.fluxTracker.setOverloadDuration(MAX_OVERLOAD_TIME)
    }

  }

  override fun getDescriptionParam(index: Int, hullSize: HullSize): String? {
    return null
  }

  override fun shouldAddDescriptionToTooltip(hullSize: HullSize, ship: ShipAPI?, isForModSpec: Boolean): Boolean {
    return true
  }

  override fun addPostDescriptionSection(tooltip: TooltipMakerAPI, hullSize: HullSize, ship: ShipAPI?, width: Float, isForModSpec: Boolean) {
    val faction = Global.getSector().getFaction(aEP_ID.FACTION_ID_FSF)
    val highLight = Misc.getHighlightColor()
    val grayColor = Misc.getGrayColor()
    val txtColor = Misc.getTextColor()
    val barBgColor = faction.getDarkUIColor()
    val factionColor: Color = faction.getBaseUIColor()
    val titleTextColor: Color = faction.getColor()

    //主效果
    tooltip.addSectionHeading(aEP_DataTool.txt("effect"), Alignment.MID, 5f)
    tooltip.addPara("{%s}"+txt("overload_time_reduce")+"{%s}", 5f, arrayOf(Color.green), aEP_ID.HULLMOD_POINT, String.format("%.0f", (OVERLOAD_TIME_DECREASE * 100f))+ "%")
    //灰字额外说明
    tooltip.addPara(aEP_DataTool.txt("MD_des04"), grayColor, 5f)
  }


  internal inner class FluxRecorder(private val ship: ShipAPI, val fluxData: floatDataRecorder) : DamageListener, DamageTakenModifier, AdvanceableListener {
    private var param: Any? = null
    private var damage: DamageAPI? = null

    override fun advance(amount: Float) {
      val maxFlux = ship.fluxTracker.maxFlux
      val softFlux = ship.fluxTracker.currFlux - ship.fluxTracker.hardFlux
      val totalDiss = aEP_Tool.getRealDissipation(ship)

      //更新上一帧幅能变化
      var fluxDecreaseCompareToLastFrame = 0f
      if(softFlux < fluxData.last){
        fluxDecreaseCompareToLastFrame = fluxData.last - softFlux
      }
      fluxData.addRenewData(softFlux * INCREASE_SPEED)

      //修正由于循环幅散和先进幅散带来的问题
      //先进幅散想停在预热状态，但是加成的幅散会更快退出预热
      val bonusFromSoft = (ship.mutableStats.fluxDissipation.getFlatStatMod(aEP_SoftfluxDissipate.ID_B)?.value?:0f) +
          (ship.mutableStats.fluxDissipation.getFlatStatMod(aEP_SoftfluxDissipate.ID_P)?.value?:0f)
      //循环幅散不想停在预热状态，但减少的幅散又会更慢退出预热
      val bonusFromBurst=(ship.mutableStats.fluxDissipation.getFlatStatMod(aEP_BurstDissipate.ID_B)?.value?:0f) +
          (ship.mutableStats.fluxDissipation.getFlatStatMod(aEP_BurstDissipate.ID_P)?.value?:0f)

      //计算缓冲区最大容量和下降速度
      val decrease =  (totalDiss * DECREASE_SPEED  - bonusFromSoft - bonusFromBurst)* amount
      val max = maxFlux * BUFFER_AREA;

      //如果幅能开始下降，持续消减buff区
      fluxData.total = MathUtils.clamp(fluxData.total - fluxDecreaseCompareToLastFrame * DECREASE_PERCENT - decrease, 0f,max)
      val buffer = fluxData.total

      //过载立刻清空预热状态
      if (ship.fluxTracker.isOverloadedOrVenting) fluxData.total = 0f
      var bufferLevel = MathUtils.clamp(buffer * 2f / (max + 1f), 0f, 1f)
      if (Global.getCombatEngine().playerShip === ship) {
        Global.getCombatEngine().maintainStatusForPlayerShip(
          this.javaClass.simpleName,  //key
          "graphics/aEP_hullsys/marker_dissipation.png",  //sprite name,full, must be registed in setting first
          Global.getSettings().getHullModSpec(ID).displayName,  //title
          aEP_DataTool.txt("MD_des01") + (bufferLevel * 100).toInt() + "%",  //data
          false
        ) //is debuff
      }
    }

    //在 report之前被调用
    //用来记录 DamageAPI
    override fun modifyDamageTaken(param: Any?, target: CombatEntityAPI, damage: DamageAPI, point: Vector2f, shieldHit: Boolean): String? {
      this.param = param
      this.damage = damage
      return null
    }

    override fun reportDamageApplied(source: Any?, target: CombatEntityAPI, result: ApplyDamageResultAPI) {
      //反向消除由于受到伤害导致的预热缓冲区变化
      if ((param is BeamAPI && damage?.isForceHardFlux == false) || damage?.isSoftFlux == true) {
        fluxData.setLastFrameData(fluxData.last + result.damageToShields)
      }
      return
    }
  }

}